File: pages/AutomationAuditPage.ts
import { type Locator } from '@playwright/test';
import { BasePage } from './BasePage';

/**
 * The four rule categories in the Automation Audit section.
 * Values match the exact heading text rendered in the app (/config/guardrails).
 *
 * NOTE: Headings are plain div/p elements — NOT semantic h1-h6 — so we use
 * getByText() to locate them, not getByRole('heading').
 */
export type AuditSection =
  | 'Emails patterns to unassign'
  | 'Subjects'
  | 'Words in User Message'
  | "Words in Assistant's Reply";

/**
 * Page Object for the Automation Audit configuration section.
 * URL: /config/guardrails
 *
 * The section lives at the BOTTOM of a long page — escalation settings render
 * first, then the Automation Audit block below. Every method scrolls the
 * target card into view before interacting.
 *
 * Each sub-section has:
 *   - A plain-text heading (NOT a semantic heading element)
 *   - A chip/tag container holding existing entries
 *   - A hidden <textarea> that receives keyboard input when the container is clicked
 *
 * Selector strategy: anchor on heading text via getByText(), scope to the
 * nearest div that also contains a textbox. Never use styled-component hashes.
 */
export class AutomationAuditPage extends BasePage {
  async navigateTo(): Promise<void> {
    await super.navigateTo('/config/guardrails');
  }

  /**
   * Returns a locator scoped to the section card that contains the given heading.
   *
   * Strategy: find all divs that contain BOTH the section heading text AND a
   * <textarea> (the tag input). .last() picks the most-nested div that satisfies
   * both conditions — the card itself, not a parent wrapper.
   */
  private sectionCard(title: AuditSection): Locator {
    return this.page
      .locator('div')
      .filter({ has: this.page.getByText(title, { exact: false }) })
      .filter({ has: this.page.getByRole('textbox') })
      .last();
  }

  /**
   * Adds an entry to the given section and waits for it to appear as a chip.
   *
   * The chip container overlays a hidden <textarea>. Clicking the container
   * focuses the textarea; typing then appends to it. Enter commits.
   */
  async addEntry(section: AuditSection, value: string): Promise<void> {
    const card = this.sectionCard(section);
    await card.scrollIntoViewIfNeeded();

    const input = card.getByRole('textbox');
    await input.click();
    await input.fill(value);
    await input.press('Enter');

    // exact: false — chip element textContent includes "×" as a child
    await card.getByText(value, { exact: false }).first().waitFor({ state: 'visible', timeout: 8_000 });
  }

  /**
   * Removes a single entry from the given section.
   * Safe to call in afterEach even if the entry was never added.
   */
  async removeEntry(section: AuditSection, value: string): Promise<void> {
    const card = this.sectionCard(section);
    await card.scrollIntoViewIfNeeded();

    // exact: false — chip element textContent includes "×" as a child element
    const chipText = card.getByText(value, { exact: false }).first();
    if ((await chipText.count()) === 0) return;

    const chipWrapper = chipText.locator('..');
    const deleteBtn = chipWrapper
      .locator('button, [role="button"], svg, span')
      .last();

    if ((await deleteBtn.count()) > 0) {
      await deleteBtn.click();
    } else {
      // Fallback: focus textarea and Backspace to remove last chip
      await card.getByRole('textbox').click();
      await card.getByRole('textbox').press('Backspace');
    }

    await chipText.waitFor({ state: 'detached', timeout: 8_000 }).catch(() => {});
  }

  /**
   * Persists pending config changes to the backend.
   *
   * After any add/remove, the page shows a "Discard / Save" toolbar.
   * Clicking Save submits the change. The toolbar disappears on success.
   *
   * Call this when tests need the rule to be active in the AI pipeline
   * (e.g. before running a Playground simulation).
   */
  async save(): Promise<void> {
    const saveBtn = this.page.getByRole('button', { name: /^save$/i });
    await saveBtn.click();
    // Toolbar disappears once the save completes
    await saveBtn.waitFor({ state: 'detached', timeout: 10_000 }).catch(() => {});
    await this.page.waitForLoadState('networkidle');
  }

  /**
   * Returns the text of all chip entries currently listed in the given section.
   */
  async getEntries(section: AuditSection): Promise<string[]> {
    const card = this.sectionCard(section);
    await card.scrollIntoViewIfNeeded();

    const raw = await card.innerText();
    const descriptionPatterns = [
      /^The AI will/i,
      /^Unassigns/i,
      /^Add a/i,
      /^Add an/i,
    ];

    return raw
      .split('\n')
      .map(t => t.replace(/×/g, '').trim())
      .filter(t => {
        if (!t || t === section) return false;
        return !descriptionPatterns.some(re => re.test(t));
      });
  }
}

File: ui/pages/automation_audit_page.py
from typing import Literal

import allure
from playwright.sync_api import Locator

from ui.pages.base_page import BasePage

# Values match the exact heading text rendered in the app (/config/guardrails).
# Headings are plain div/p elements — NOT semantic h1-h6 — so we use
# get_by_text(), not get_by_role("heading").
AuditSection = Literal[
    "Emails patterns to unassign",
    "Subjects",
    "Words in User Message",
    "Words in Assistant's Reply",
]


class AutomationAuditPage(BasePage):
    @allure.step("Navigate to Guardrails config")
    def navigate_to(self) -> None:
        from config import TIMEOUTS
        super().navigate_to("/config/guardrails")
        # Wait for the guardrails content to render (SPA fetches data after navigation).
        # "Emails patterns" is always the first visible section heading on this page.
        self.page.get_by_text("Emails patterns", exact=False).wait_for(
            state="visible", timeout=TIMEOUTS["element"]
        )

    def _section_card(self, title: AuditSection) -> Locator:
        return (
            self.page.locator("div")
            .filter(has=self.page.get_by_text(title, exact=False))
            .filter(has=self.page.get_by_role("textbox"))
            .last
        )

    @allure.step("Add '{value}' to {section}")
    def add_entry(self, section: AuditSection, value: str) -> None:
        card = self._section_card(section)
        card.scroll_into_view_if_needed()

        inp = card.get_by_role("textbox")
        inp.click()
        inp.fill(value)
        inp.press("Enter")

        # exact=False — chip textContent includes "×" as a child element
        card.get_by_text(value, exact=False).first.wait_for(state="visible", timeout=8_000)

    @allure.step("Remove '{value}' from {section}")
    def remove_entry(self, section: AuditSection, value: str) -> None:
        card = self._section_card(section)
        card.scroll_into_view_if_needed()

        chip_text = card.get_by_text(value, exact=False).first
        if chip_text.count() == 0:
            return

        # Chip wrapper contains the label + a close affordance (×).
        # Go up to the wrapper and click the last child (the × button).
        chip_wrapper = chip_text.locator("..")
        delete_btn = chip_wrapper.locator('button, [role="button"], svg, span').last

        if delete_btn.count() > 0:
            delete_btn.click()
        else:
            # Fallback: focus textarea and Backspace to remove last chip
            card.get_by_role("textbox").click()
            card.get_by_role("textbox").press("Backspace")

        chip_text.wait_for(state="detached", timeout=8_000)

    @allure.step("Save config changes")
    def save(self) -> None:
        # Use data-testid to avoid strict-mode violation when multiple "Save" buttons exist.
        save_btn = self.page.get_by_test_id("config-save-button")
        if save_btn.count() == 0:
            return  # no pending changes — toolbar not visible
        save_btn.click()

        # A confirmation dialog appears after clicking Save.
        # Click the dialog's "Save" button to confirm (distinct from the toolbar button).
        confirm_save = self.page.locator("#page-content").get_by_role("button", name="Save")
        try:
            confirm_save.wait_for(state="visible", timeout=5_000)
            confirm_save.click()
        except Exception:
            pass  # no confirmation dialog — save went through directly

        # Wait for the toolbar (Discard/Save) to disappear — indicates save completed.
        save_btn.wait_for(state="detached", timeout=30_000)
        self.page.wait_for_load_state("load")

